src/jit/decomposelongs.cpp

```diff
@@ -1030,6 +1030,51 @@ GenTree* DecomposeLongs::DecomposeMul(LIR::Use& use)
 }
 
 //------------------------------------------------------------------------
+// DecomposeUMod: Decompose GT_UMOD. The only GT_UMODs that make it to decompose
+// are guaranteed to be an unsigned long mod with op2 which is a cast to long from
+// a constant int whose value is between 2 and 0x3fffffff. All other GT_UMODs are
+// morphed into helper calls. These GT_UMODs will actually return an int value in
+// RDX. In decompose, we change the type of the operation to a TYP_INT, and replace
+// op2 with its low half.
+//
+// Arguments:
+//    use - the LIR::Use object for the def that needs to be decomposed.
+//
+// Return Value:
+//    The next node to process.
+//
+GenTree* DecomposeLongs::DecomposeUMod(LIR::Use& use)
+{
+    assert(use.IsInitialized());
+
+    GenTree*   tree = use.Def();
+    genTreeOps oper = tree->OperGet();
+
+    assert(oper == GT_UMOD);
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+    assert(op1->OperGet() == GT_LONG);
+    assert(op2->OperGet() == GT_LONG);
+
+    GenTree* loOp2 = op2->gtGetOp1();
+    GenTree* hiOp2 = op2->gtGetOp2();
+
+    assert(loOp2->OperGet() == GT_CNS_INT);
+    assert(hiOp2->OperGet() == GT_CNS_INT);
+    assert(loOp2->gtIntCon.gtIconVal >= 2 && loOp2->gtIntCon.gtIconVal <= 0x3fffffff && hiOp2->gtIntCon.gtIconVal == 0);
```

Comment:
I would suggest to have 2 separate asserts instead of clubbing them into one so that when an assert fails we know the specific condition that failed without requiring to put under debugger.

e.g
```
assert(loOp2->gtIntCon.gtIconVal >= 2 && loOp2->gtIntCon.gtIconVal <= 0x3fffffff );
assert(hiOp2->gtIntCon.gtIconVal == 0);
```